server/src/main/java/org/apache/druid/segment/realtime/appenderator/BaseAppenderatorDriver.java

```diff
@@ -675,12 +676,19 @@ ListenableFuture<SegmentsAndCommitMetadata> publishInBackground(
                 segmentsAndCommitMetadata.getSegments(),
                 "Failed publish, not removing segments"
             );
-            Throwables.propagateIfPossible(e);
-            throw new RuntimeException(e);
+            if (e.getMessage() != null && e.getMessage().contains("Failed to update the metadata Store. The new start metadata is ahead of last commited end state.")) {
+              // we can recover from this error.
+              throw new ExecutionException(e);
```

Comment:
Why do we need to throw an ExecutionException instead of checking the message itself below?